File: find_cross_imports.py

import os
import re

features_dir = r'c:\Users\agnes\Desktop\projects\tengra\src\renderer\features'
features = [d for d in os.listdir(features_dir) if os.path.isdir(os.path.join(features_dir, d))]

import_re = re.compile(r'from [\'"]@/features/([^/\'"]+)')

cross_imports = {}

for feature in features:
    feature_path = os.path.join(features_dir, feature)
    for root, dirs, files in os.walk(feature_path):
        for file in files:
            if file.endswith('.tsx') or file.endswith('.ts'):
                file_path = os.path.join(root, file)
                try:
                    with open(file_path, 'r', encoding='utf-8') as f:
                        content = f.read()
                        matches = import_re.findall(content)
                        for match in matches:
                            if match != feature:
                                if match not in cross_imports:
                                    cross_imports[match] = set()
                                cross_imports[match].add(feature)
                                print(f"File {file_path} imports from {match}")
                except Exception as e:
                    print(f"Error reading {file_path}: {e}")

print("\nCross-feature import summary:")
for target, sourcers in cross_imports.items():
    print(f"Feature '{target}' is imported by: {', '.join(sourcers)}")

File: find_cross_imports.js

const fs = require('fs');
const path = require('path');

const featuresDir = 'c:\\Users\\agnes\\Desktop\\projects\\tengra\\src\\renderer\\features';
const features = fs.readdirSync(featuresDir).filter(f => fs.statSync(path.join(featuresDir, f)).isDirectory());

const importRe = /from ['"]@\/features\/([^/'"]+)/g;

const crossImports = {};

function walk(dir, callback) {
    fs.readdirSync(dir).forEach(f => {
        const fullPath = path.join(dir, f);
        if (fs.statSync(fullPath).isDirectory()) {
            walk(fullPath, callback);
        } else {
            callback(fullPath);
        }
    });
}

features.forEach(feature => {
    const featurePath = path.join(featuresDir, feature);
    walk(featurePath, (filePath) => {
        if (filePath.endsWith('.tsx') || filePath.endsWith('.ts')) {
            try {
                const content = fs.readFileSync(filePath, 'utf-8');
                let match;
                while ((match = importRe.exec(content)) !== null) {
                    const importedFeature = match[1];
                    if (importedFeature !== feature) {
                        if (!crossImports[importedFeature]) {
                            crossImports[importedFeature] = new Set();
                        }
                        crossImports[importedFeature].add(feature);
                        console.log(`File ${filePath} imports from ${importedFeature}`);
                    }
                }
            } catch (e) {
                console.error(`Error reading ${filePath}: ${e}`);
            }
        }
    });
});

console.log("\nCross-feature import summary:");
for (const [target, sourcers] of Object.entries(crossImports)) {
    console.log(`Feature '${target}' is imported by: ${Array.from(sourcers).join(', ')}`);
}
